rclpy/test/test_node.py

```diff
@@ -1212,6 +1238,184 @@ def test_set_descriptor_read_only(self):
+            ('int_value', 123, ParameterDescriptor(floating_point_range=[fp_range]))
+        ]
+
+        result = self.node.declare_parameters('', parameters)
+
+        self.assertIsInstance(result, list)
+        self.assertIsInstance(result[0], Parameter)
+        self.assertIsInstance(result[1], Parameter)
+        self.assertIsInstance(result[2], Parameter)
+        self.assertIsInstance(result[3], Parameter)
+        self.assertIsInstance(result[4], Parameter)
+        self.assertAlmostEqual(result[0].value, 0.0)
+        self.assertAlmostEqual(result[1].value, 10.0)
+        self.assertAlmostEqual(result[2].value, 4.5)
+        self.assertEqual(result[3].value, 'I am no integer')
+        self.assertEqual(result[4].value, 123)
+        self.assertEqual(self.node.get_parameter('from_value').value, 0.0)
+        self.assertEqual(self.node.get_parameter('to_value').value, 10.0)
+        self.assertEqual(self.node.get_parameter('in_range').value, 4.5)
+        self.assertEqual(self.node.get_parameter('str_value').value, 'I am no integer')
+        self.assertAlmostEqual(self.node.get_parameter('int_value').value, 123)
+
+        # Try to set a parameter out of range.
+        result = self.node.set_parameters([Parameter('in_range', value=12.0)])
+        self.assertIsInstance(result, list)
+        self.assertIsInstance(result[0], SetParametersResult)
+        self.assertFalse(result[0].successful)
+        self.assertEqual(self.node.get_parameter('in_range').value, 4.5)
+
+        # Try to set a parameter out of range (bad step).
+        result = self.node.set_parameters([Parameter('in_range', value=4.25)])
+        self.assertIsInstance(result, list)
+        self.assertIsInstance(result[0], SetParametersResult)
+        self.assertFalse(result[0].successful)
+        self.assertEqual(self.node.get_parameter('in_range').value, 4.5)
+
+        # Change in_range parameter to int; ranges will not apply.
+        result = self.node.set_parameters([Parameter('in_range', value=12)])
+        self.assertIsInstance(result, list)
+        self.assertIsInstance(result[0], SetParametersResult)
+        self.assertTrue(result[0].successful)
+        self.assertEqual(self.node.get_parameter('in_range').value, 12)
+
+        # From and to are always valid.
+        # Parameters that don't comply with the description will raise an exception.
+        fp_range = FloatingPointRange(from_value=-10.0, to_value=0.0, step=30.0)
+        parameters = [
+            ('from_value_2', -10.0, ParameterDescriptor(floating_point_range=[fp_range])),
+            ('to_value_2', 0.0, ParameterDescriptor(floating_point_range=[fp_range])),
+            ('in_range_bad_step', -4.5, ParameterDescriptor(floating_point_range=[fp_range])),
+            ('out_of_range', 30.0, ParameterDescriptor(floating_point_range=[fp_range]))
+        ]
+        with self.assertRaises(InvalidParameterValueException):
+            self.node.declare_parameters('', parameters)
```

Comment:
Should `declare_parameters` raise, or simply return a not successful result (in the corresponding list item).
Doing both things sounds strange to me.

Reply:
Declaring parameters is something you should do once in your program, and as a user you should probably know the constraints and the initial value beforehand. Following that line, I think it makes sense to raise an exception.

This is the case in `rclcpp` AFAIK: https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/node.hpp#L390-L391.

On the other hand, parameters are declared in order. If the first ones in the list are correct, they will be declared until the 'faulty' one is reached. That part is explicitly documented like that.